src/lib.rs

```diff
@@ -6,177 +56,258 @@ extern crate base64;
+///
+/// Fields:
+/// 
+/// - url: where to find the corrsponding file online
+/// - name: name of the corrsponding file
+/// - size: of the file
+#[derive(Debug, Clone)]
+pub struct DlcLink {
     pub url: String,
     pub name: String,
     pub size: String
 }
 
-impl FileData {
-    fn new() -> FileData {
-        FileData {
+impl DlcLink {
+    fn new() -> DlcLink {
+        DlcLink {
             url: String::new(),
             name: String::new(),
             size: String::new(),
         }
     }
 }
 
-#[derive(Debug)]
-pub struct PkgData {
+/// The readable result of an .dlc file. 
+/// 
+/// It holds the information about:
+/// 
+/// - name: of the .dlc package
+/// - password: ??? pwd to encrypt ???
+/// - files: the corresponding files of the package
+#[derive(Debug, Clone)]
+pub struct DlcPackage {
     pub name: String,
-    pub pwd: String,
-    pub files: Vec<FileData>,
+    pub password: String,
+    pub files: Vec<DlcLink>,
 }
 
-impl PkgData {
-    fn new() -> PkgData {
-        PkgData {
+impl DlcPackage {
+    fn new() -> DlcPackage {
+        DlcPackage {
             name: String::new(),
-            pwd: String::new(),
+            password: String::new(),
             files: Vec::new(),
         }
     }
 }
 
+/// Struct to decode the .dlc file or data into an readable format.
+#[derive(Debug)]
+pub struct DlcDecoder {
+    jd_app_name: String,
+    jd_decryption_key: Vec<u8>,
+    jd_decryption_iv: Vec<u8>
+}
 
-pub fn decrypt_dlc(data: Vec<u8>, app_name: &str, dec_key: &[u8], dec_iv: &[u8]) -> Result<PkgData> {
-    let data = decrypt_raw(data, app_name, dec_key, dec_iv)?;
-    let data = String::from_utf8(data)?;
-    let mut pkg = PkgData::new();
-
-    let re = Regex::new(r#"<package ([^>]*)"#)?;
-    let t = re.find(&data).ok_or("Can't find package in data")?;
-    let (name, pwd) = match pkg_details(t.as_str().to_string()) {
-        Ok((name, pwd)) => (name, pwd),
-        Err(_)          => (String::new(), String::new()),
-    };
-
-    pkg.name = name;
-    pkg.pwd = pwd;
-
-    let files_xml: Vec<&str> = data.split("<file>").collect();
-    for f in files_xml {
-        let details: Vec<&str> = f.split("<").collect();
-        let mut file = FileData::new();
-        for d in details {
-            if d.len() > 3 && d[..3] == "url".to_string() {
-                let buf = file_details(d.to_string(), 4);
-                file.url = buf;
-            } else if d.len() > 8 && d[..8] == "filename".to_string() {
-                let buf = file_details(d.to_string(), 9);
-                file.name = buf;
-            } else if d.len() > 4 && d[..4] == "size".to_string() {
-                let buf = file_details(d.to_string(), 5);
-                file.size = buf;
-            }
+impl DlcDecoder {
+    /// Create a new DlcDecoder with a standard login to jdownloader.
+    pub fn new() -> DlcDecoder {
+        DlcDecoder {
+            jd_app_name: "pylo".to_string(),
```

Comment:
The jd_app_name, jd_decryption_key and jd_decryption_iv must be setable from the program which uses the library, because all 3 information are taken from pyload. The jdownloader group can ban this combination if they like.

Reply:
You are totally right!
What about the following change? I added a function to change the credentials and keeper the pyload as default fallback.

This would allow us to use both use-cases:
1. Easy extraction of the .dlc container without the knowledge behind it
2. Enhanced extraction with the own credentials, for these who have them/knew it

Reply:
Thats great. Thank you.